src/compute-client/src/explain.rs

```diff
@@ -253,22 +257,105 @@ pub struct TimestampSource<T> {
     pub write_frontier: Vec<T>,
 }
 
-impl<T: fmt::Display + fmt::Debug> fmt::Display for TimestampExplanation<T> {
+pub trait DisplayableInTimeline {
+    fn fmt(&self, timeline: Option<&Timeline>, f: &mut fmt::Formatter) -> fmt::Result;
+    fn display<'a>(&'a self, timeline: Option<&'a Timeline>) -> DisplayInTimeline<'a, Self> {
+        DisplayInTimeline { t: self, timeline }
+    }
+}
+
+impl DisplayableInTimeline for mz_repr::Timestamp {
+    fn fmt(&self, timeline: Option<&Timeline>, f: &mut fmt::Formatter) -> fmt::Result {
+        match timeline {
+            Some(Timeline::EpochMilliseconds) => {
+                let ts_ms: u64 = self.into();
+                let ts = ts_ms / 1000;
+                let nanos = ((ts_ms % 1000) as u32) * 1000000;
+                let ndt = NaiveDateTime::from_timestamp(ts as i64, nanos);
+                write!(f, "{:13} ({})", self, ndt.format("%Y-%m-%d %H:%M:%S%.3f"))
+            }
+            None | Some(_) => {
+                write!(f, "{:13}", self)
+            }
+        }
+    }
+}
+
+pub struct DisplayInTimeline<'a, T: ?Sized> {
+    t: &'a T,
+    timeline: Option<&'a Timeline>,
+}
+impl<'a, T> fmt::Display for DisplayInTimeline<'a, T>
+where
+    T: DisplayableInTimeline,
+{
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        writeln!(f, "          query timestamp: {:13}", self.timestamp)?;
-        writeln!(f, "                    since:{:13?}", self.since)?;
-        writeln!(f, "                    upper:{:13?}", self.upper)?;
+        self.t.fmt(self.timeline, f)
+    }
+}
+
+impl<'a, T> fmt::Debug for DisplayInTimeline<'a, T>
+where
+    T: DisplayableInTimeline,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self, f)
+    }
+}
+
+impl<T: fmt::Display + fmt::Debug + DisplayableInTimeline> fmt::Display
+    for TimestampExplanation<T>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let timeline = self.timeline.as_ref();
+        writeln!(
+            f,
+            "          query timestamp: {}",
+            self.timestamp.display(timeline)
+        )?;
+        writeln!(
+            f,
+            "          since:{:?}",
```

Comment:
The columns are no longer aligned. Previously all timestamps were exactly aligned so it was as easy as possible to visually check them. We should preserve that property.